colossalai/shardformer/layer/loss.py

```diff
@@ -78,17 +78,21 @@ def forward(ctx, vocab_logits: torch.Tensor, target: torch.Tensor, ignore_index:
         # calculate the loss
         # loss = log(sum(exp(x[i]))) - x[class]
         loss = torch.where(target == ignore_index, 0.0, torch.log(sum_exp_logits) - pred_logits)
-        loss = torch.sum(loss).div_(torch.sum(loss != 0.0))
+        non_zero_sum = torch.sum(loss != 0.0)
+        ctx.mean_grad = 1.0 / non_zero_sum
+        loss = torch.sum(loss).div_(non_zero_sum)
```

Comment:
What does this mean? loss = sum * num_no_zero?

Reply:
I calculate the loss by loss = sum / num_no_zero.